x/warp/contract.go

```diff
@@ -240,46 +231,33 @@ func UnpackGetVerifiedWarpMessageOutput(output []byte) (GetVerifiedWarpMessageOu
 	return outputStruct, err
 }
 
-func unpackWarpMessage(accessibleState contract.AccessibleState, input []byte) ([]byte, bool) {
+type messageHandler interface {
+	packFailed() []byte
+	handleMessage(msg *warp.Message) ([]byte, error)
+}
+
+func handleWarpMessage(accessibleState contract.AccessibleState, input []byte, remainingGas uint64, handler messageHandler) ([]byte, uint64, error) {
 	warpIndex, err := UnpackGetVerifiedWarpMessageInput(input)
 	if err != nil {
-		return nil, false
+		return nil, 0, fmt.Errorf("%w: %s", errInvalidIndexInput, err)
 	}
 	if !warpIndex.IsInt64() {
-		return nil, false
+		return nil, 0, fmt.Errorf("%w: %v", errInvalidIndexInput, warpIndex)
 	}
 	warpIndexInt := int(warpIndex.Int64())
 	state := accessibleState.GetStateDB()
 	predicateBytes, exists := state.GetPredicateStorageSlots(ContractAddress, warpIndexInt)
 	predicateResults := accessibleState.GetBlockContext().GetPredicateResults(state.GetTxHash(), ContractAddress)
-	valid := set.BitsFromBytes(predicateResults).Contains(warpIndexInt)
-	return predicateBytes, exists && valid
-}
-
-// getVerifiedWarpMessage retrieves the pre-verified warp message from the predicate storage slots and returns
-// the expected ABI encoding of the message to the caller.
-func getVerifiedWarpMessage(accessibleState contract.AccessibleState, caller common.Address, addr common.Address, input []byte, suppliedGas uint64, readOnly bool) (ret []byte, remainingGas uint64, err error) {
-	remainingGas, err = contract.DeductGas(suppliedGas, GetVerifiedWarpMessageBaseCost)
-	if err != nil {
-		return nil, remainingGas, err
-	}
-	predicateBytes, valid := unpackWarpMessage(accessibleState, input)
-	// If there is no such value or it failed verification, return invalid
+	valid := set.BitsFromBytes(predicateResults).Contains(warpIndexInt) && exists
 	if !valid {
-		packedOutput, err := PackGetVerifiedWarpMessageOutput(GetVerifiedWarpMessageOutput{
-			Valid: false,
-		})
-		if err != nil {
-			return nil, remainingGas, err
-		}
-		return packedOutput, remainingGas, nil
+		return handler.packFailed(), remainingGas, nil
 	}
 
 	// Note: we charge for the size of the message during both predicate verification and each time the message is read during
 	// EVM execution because each execution incurs an additional read cost.
 	msgBytesGas, overflow := math.SafeMul(GasCostPerWarpMessageBytes, uint64(len(predicateBytes)))
 	if overflow {
-		return nil, remainingGas, vmerrs.ErrOutOfGas
+		return nil, 0, vmerrs.ErrOutOfGas
```

Comment:
in template we don't follow this pattern, instead we return remaining gas. Should we follow this pattern here as well?

Reply:
Where in the template do we return `remainingGas` alongside `ErrOutOfGas` ? I don't seem to see it and think it seems weird to return non-zero gas with `ErrOutOfGas`. I think we may do this for all other errors?